packages/time-oracle/src/migrate_v1.rs

```diff
@@ -0,0 +1,158 @@
+                (vec![], vec![]),
+                |mut v: (Vec<AlarmOld>, Vec<Id>), alarm| -> Result<_, StdError> {
+                    let alarm = alarm?;
+                    v.1.push(alarm.0);
+                    if let Some(last) = v.0.last() {
+                        if last.addr != alarm.1.addr {
+                            v.0.push(alarm.1)
+                        }
+                    } else {
+                        v.0.push(alarm.1)
+                    }
+                    Ok(v)
+                },
+            )?;
+
+        // purge all the data
+        for id in ids {
+            old_alarms.remove(storage, id)?;
+        }
+        self.next_id.remove(storage);
+
+        let alarms_new = Alarms::new(self.namespace_alarms, self.namespace_index);
+
+        // restore to new alarms
+        for alarm in alarms {
+            alarms_new.add(storage, alarm.addr, Timestamp::from_seconds(alarm.time))?;
+        }
+
+        Ok(())
+    }
+
+    #[cfg(test)]
+    fn add(
+        &self,
+        storage: &mut dyn Storage,
+        addr: Addr,
+        time: Timestamp,
+    ) -> Result<(), AlarmError> {
+        let id = self.next_id.may_load(storage)?.unwrap_or_default();
+
+        let alarm = AlarmOld {
+            time: as_seconds(time),
+            addr,
+        };
+
+        self.alarms().save(storage, id, &alarm)?;
+
+        self.next_id.save(storage, &id.wrapping_add(1))?;
+
+        Ok(())
+    }
+
+    fn alarms(&self) -> IndexedMap<'a, TimeSeconds, AlarmOld, AlarmIndexes<'a>> {
+        let indexes = AlarmIndexes {
+            alarms: MultiIndex::new(|_, d| d.time, self.namespace_alarms, self.namespace_index),
+        };
+
+        IndexedMap::new(self.namespace_alarms, indexes)
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use sdk::cosmwasm_std::testing;
+
+    use super::*;
+
+    #[test]
+    fn test_migration() {
+        let alarms = AlarmsOld::new("alarms", "alarms_idx", "alarms_next_id");
+        let storage = &mut testing::mock_dependencies().storage;
+        let t1 = Timestamp::from_seconds(1);
+        let t2 = Timestamp::from_seconds(2);
+        let t3 = Timestamp::from_seconds(3);
+        let t4 = Timestamp::from_seconds(4);
+        let addr1 = Addr::unchecked("addr1");
+        let addr2 = Addr::unchecked("addr2");
+        let addr3 = Addr::unchecked("addr3");
+        let addr4 = Addr::unchecked("addr4");
+
+        alarms.add(storage, addr1, t1).unwrap();
+        alarms.add(storage, addr2, t1).unwrap();
+        alarms.add(storage, addr3.clone(), t2).unwrap();
+        alarms.add(storage, addr3, t3).unwrap();
+        alarms.add(storage, addr4, t4).unwrap();
+
+        // multiple alarms per address(5) + index(5) + next_id(1)
+        assert_eq!(11, storage.range(None, None, Order::Ascending).count());
+
+        alarms.migrate(storage).unwrap();
+
+        // single alarm per address(4) + index(4)
+        assert_eq!(8, storage.range(None, None, Order::Ascending).count());
```

Comment:
please add more checks on the alarms post-migration, e.g. what alarms for what addresses are there